indexer/api/api.go

```diff
@@ -3,43 +3,114 @@ package api
 	"github.com/go-chi/chi/v5"
 	"github.com/go-chi/chi/v5/middleware"
+	"github.com/prometheus/client_golang/prometheus"
 )
 
 const ethereumAddressRegex = `^0x[a-fA-F0-9]{40}$`
 
 type Api struct {
-	log    log.Logger
-	Router *chi.Mux
+	log             log.Logger
+	Router          *chi.Mux
+	apiConfig       config.APIConfig
+	metricsConfig   config.MetricsConfig
+	metricsRegistry *prometheus.Registry
 }
 
-func NewApi(logger log.Logger, bv database.BridgeTransfersView) *Api {
-	r := chi.NewRouter()
-	h := routes.NewRoutes(logger, bv, r)
+const (
+	MetricsNamespace = "op_indexer"
+)
+
+func ChiMetricsMiddleware(rec metrics.HTTPRecorder) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return metrics.NewHTTPRecordingMiddleware(rec, next)
+	}
+}
+
+func NewApi(logger log.Logger, bv database.BridgeTransfersView, apiConfig config.APIConfig, metricsConfig config.MetricsConfig) *Api {
+	apiRouter := chi.NewRouter()
+	h := routes.NewRoutes(logger, bv, apiRouter)
+
+	mr := metrics.NewRegistry()
+	promRecorder := metrics.NewPromHTTPRecorder(mr, MetricsNamespace)
+
+	apiRouter.Use(ChiMetricsMiddleware(promRecorder))
+	apiRouter.Use(middleware.Heartbeat("/healthz"))
+
+	apiRouter.Get(fmt.Sprintf("/api/v0/deposits/{address:%s}", ethereumAddressRegex), h.L1DepositsHandler)
+	apiRouter.Get(fmt.Sprintf("/api/v0/withdrawals/{address:%s}", ethereumAddressRegex), h.L2WithdrawalsHandler)
+
+	return &Api{log: logger, Router: apiRouter, metricsRegistry: mr, metricsConfig: metricsConfig, apiConfig: apiConfig}
+}
+
+func (a *Api) Start(ctx context.Context) error {
+	var wg sync.WaitGroup
+	errCh := make(chan error, 2)
+
+	processCtx, processCancel := context.WithCancel(ctx)
+	runProcess := func(start func(ctx context.Context) error) {
+		wg.Add(1)
+		go func() {
+			defer func() {
+				if err := recover(); err != nil {
+					a.log.Error("halting api on panic", "err", err)
+					debug.PrintStack()
+					errCh <- fmt.Errorf("panic: %v", err)
+				}
+
+				processCancel()
+				wg.Done()
+			}()
 
-	r.Use(middleware.Heartbeat("/healthz"))
+			errCh <- start(processCtx)
+		}()
+	}
+
+	runProcess(a.startServer)
+	runProcess(a.startMetricsServer)
+
+	wg.Wait()
+
+	err := <-errCh
+	if err != nil {
+		a.log.Error("api stopped", "err", err)
+	} else {
+		a.log.Info("api stopped")
+	}
 
-	r.Get(fmt.Sprintf("/api/v0/deposits/{address:%s}", ethereumAddressRegex), h.L1DepositsHandler)
-	r.Get(fmt.Sprintf("/api/v0/withdrawals/{address:%s}", ethereumAddressRegex), h.L2WithdrawalsHandler)
-	return &Api{log: logger, Router: r}
+	return err
 }
 
-func (a *Api) Listen(ctx context.Context, port int) error {
-	a.log.Info("api server listening...", "port", port)
-	server := http.Server{Addr: fmt.Sprintf(":%d", port), Handler: a.Router}
+func (a *Api) startServer(ctx context.Context) error {
+	a.log.Info("api server listening...", "port", a.apiConfig.Port)
+	server := http.Server{Addr: fmt.Sprintf(":%d", a.apiConfig.Port), Handler: a.Router}
```

Comment:
Prefer to pass in the port when listen is called but doesn't really matter much in this case and this allowed me to copy paste `runProcess`